### PR TITLE
mike/undef_pip_size_error_on_restarts

### PR DESCRIPTION
**Pull request by michaelmueller-binary**

### CODE REVIEW

File: lib/Mojo/WebSocketProxy/Backend/JSONRPC.pm [hunk 1]
```diff
@@ -15,22 +15,72 @@ use MojoX::JSON::RPC::Client;
 
 ## VERSION
 
-__PACKAGE__->register_type( 'jsonrpc' );
+__PACKAGE__->register_type('jsonrpc');
 
 sub url { return shift->{url} }
 
 my $request_number = 0;
+=head2 call_rpc
```
Comment (michaelmueller-binary) [inline lib/Mojo/WebSocketProxy/Backend/JSONRPC.pm]: documentation change

File: lib/Mojo/WebSocketProxy/Backend/JSONRPC.pm [hunk 2]
```diff
@@ -40,10 +90,11 @@ sub call_rpc {
     my $before_get_rpc_response_hook = delete($req_storage->{before_get_rpc_response}) || [];
     my $after_got_rpc_response_hook  = delete($req_storage->{after_got_rpc_response})  || [];
     my $before_call_hook             = delete($req_storage->{before_call})             || [];
+    my $rpc_failure_cb               =  delete($req_storage->{rpc_failure_cb}) || 0;
```
Comment (michaelmueller-binary) [inline lib/Mojo/WebSocketProxy/Backend/JSONRPC.pm]: addition of call back

File: lib/Mojo/WebSocketProxy/Backend/JSONRPC.pm [hunk 3]
```diff
@@ -52,46 +103,48 @@ sub call_rpc {
...
+
+                my $api_response;
+                if (!$res) {
+                    my $tx      = $client->tx;
+                    my $details = 'URL: ' . $tx->req->url;
+                    if (my $err = $tx->error) {
+                        $details .= ', code: ' . ($err->{code} // 'n/a') . ', response: ' . $err->{message};
+                    }
+                    warn "WrongResponse [$msg_type], details: $details";
+                    $rpc_failure_cb->($c, $res, $req_storage ) if $rpc_failure_cb;
```
Comment (michaelmueller-binary) [inline lib/Mojo/WebSocketProxy/Backend/JSONRPC.pm]: running of call back on error

File: lib/Mojo/WebSocketProxy/Backend/JSONRPC.pm [hunk 4]
```diff
@@ -52,46 +103,48 @@ sub call_rpc {
...
 
-            if ($res->is_error) {
-                warn $res->error_message;
-                $api_response = $c->wsp_error($msg_type, 'CallError', 'Sorry, an error occurred while processing your request.');
-                $c->send({json => $api_response}, $req_storage);
-                return;
-            }
+                if ($res->is_error) {
+                    warn $res->error_message;
+                    $rpc_failure_cb->($c, $res, $req_storage ) if $rpc_failure_cb;
```
Comment (michaelmueller-binary) [inline lib/Mojo/WebSocketProxy/Backend/JSONRPC.pm]: running of call back on error

File: lib/Mojo/WebSocketProxy/Backend/JSONRPC.pm [hunk 5]
```diff
@@ -15,22 +15,72 @@ use MojoX::JSON::RPC::Client;
 
 ## VERSION
 
-__PACKAGE__->register_type( 'jsonrpc' );
```
Comment (michaelmueller-binary) [inline lib/Mojo/WebSocketProxy/Backend/JSONRPC.pm]: Added some comments to identify actual changes V Perl tidy

File: lib/Mojo/WebSocketProxy/Backend/JSONRPC.pm (hunk 2 above)
Comment (chylli-binary) [reply lib/Mojo/WebSocketProxy/Backend/JSONRPC.pm]: Will `undef` be better ? or a blank sub `sub {}` ?

File: lib/Mojo/WebSocketProxy/Backend/JSONRPC.pm (hunk 1 above)
Comment (tom-binary) [inline lib/Mojo/WebSocketProxy/Backend/JSONRPC.pm]: This needs blank lines before+after the `=pod` lines.

```suggestion

=head2 call_rpc
```

**Pull request merged**